### ISSUE TITLE
Using V 2.0.0 Text file in CentOS 7 do not get decrypted properly - extra 4 bytes at the end of the decrypted file

### ISSUE THREAD

**Issue by willhanson98** [coderarjob/tea-c#3]
Using V 2.0.0 Text file in CentOS 7 do not get decrypted properly - extra 4 bytes at the end of the decrypted file.
Dump of original unencrypted:
od -bc secret1
0000000 164 150 151 163 040 151 163 040 167 150 141 164 040 151 163 040
          t   h   i   s       i   s       w   h   a   t       i   s
0000020 151 156 040 163 145 143 162 145 164 061 012 163 157 155 145 040
          i   n       s   e   c   r   e   t   1  \n   s   o   m   e
0000040 165 160 160 145 162 143 141 163 145 040 101 102 103 040 104 105
          u   p   p   e   r   c   a   s   e       A   B   C       D   E
0000060 106 040 107 110 111 040 112 113 114 115 040 116 117 120 121 122
          F       G   H   I       J   K   L   M       N   O   P   Q   R
0000100 040 123 124 125 040 126 127 040 130 131 132 012 163 157 155 145
              S   T   U       V   W       X   Y   Z  \n   s   o   m   e
0000120 040 156 165 155 142 145 162 163 040 061 062 063 040 064 065 066
              n   u   m   b   e   r   s       1   2   3       4   5   6
0000140 040 067 070 071 056 061 062 040 055 065 065 065 056 067 070 012
              7   8   9   .   1   2       -   5   5   5   .   7   8  \n
0000160 163 157 155 145 040 163 171 155 142 157 154 163 040 044 046 136
          s   o   m   e       s   y   m   b   o   l   s       $   &   ^
0000200 045 043 100 041 176 052 050 051 137 053 174 012
          %   #   @   !   ~   *   (   )   _   +   |  \n

Dump of encrypted file:
od -bc secret1.3
0000000 264 256 170 046 111 333 344 227 203 305 174 367 252 142 117 330
        264 256   x   &   I 333 344 227 203 305   | 367 252   b   O 330
0000020 241 301 351 307 005 033 211 322 207 022 212 064 377 352 365 254
        241 301 351 307 005 033 211 322 207 022 212   4 377 352 365 254
0000040 062 311 032 273 126 242 131 326 256 157 135 265 005 025 222 305
          2 311 032 273   V 242   Y 326 256   o   ] 265 005 025 222 305
0000060 272 360 275 165 177 357 236 357 030 302 020 235 265 317 242 301
        272 360 275   u 177 357 236 357 030 302 020 235 265 317 242 301
0000100 262 236 010 270 144 104 143 060 162 241 334 010 026 000 136 210
        262 236  \b 270   d   D   c   0   r 241 334  \b 026  \0   ^ 210
0000120 203 050 175 347 337 252 323 210 360 365 255 157 143 053 253 345
        203   (   } 347 337 252 323 210 360 365 255   o   c   + 253 345
0000140 336 271 021 115 345 046 016 057 164 135 070 106 322 226 173 170
        336 271 021   M 345   & 016   /   t   ]   8   F 322 226   {   x
0000160 225 250 371 161 122 263 032 045 344 107 075 312 030 010 314 232
        225 250 371   q   R 263 032   % 344   G   = 312 030  \b 314 232
0000200 230 054 257 121 121 022 257 013 056 350 335 137 120 314 120 051
        230   , 257   Q   Q 022 257  \v   . 350 335   _   P 314   P   )
0000220

Dump of decrypted file:
od -bc secret1
0000000 164 150 151 163 040 151 163 040 167 150 141 164 040 151 163 040
          t   h   i   s       i   s       w   h   a   t       i   s
0000020 151 156 040 163 145 143 162 145 164 061 012 163 157 155 145 040
          i   n       s   e   c   r   e   t   1  \n   s   o   m   e
0000040 165 160 160 145 162 143 141 163 145 040 101 102 103 040 104 105
          u   p   p   e   r   c   a   s   e       A   B   C       D   E
0000060 106 040 107 110 111 040 112 113 114 115 040 116 117 120 121 122
          F       G   H   I       J   K   L   M       N   O   P   Q   R
0000100 040 123 124 125 040 126 127 040 130 131 132 012 163 157 155 145
              S   T   U       V   W       X   Y   Z  \n   s   o   m   e
0000120 040 156 165 155 142 145 162 163 040 061 062 063 040 064 065 066
              n   u   m   b   e   r   s       1   2   3       4   5   6
0000140 040 067 070 071 056 061 062 040 055 065 065 065 056 067 070 012
              7   8   9   .   1   2       -   5   5   5   .   7   8  \n
0000160 163 157 155 145 040 163 171 155 142 157 154 163 040 044 046 136
          s   o   m   e       s   y   m   b   o   l   s       $   &   ^
0000200 045 043 100 041 176 052 050 051 137 053 174 012 000 000 000 000
          %   #   @   !   ~   *   (   )   _   +   |  \n  \0  \0  \0  \0
0000220

Ran this to encrypt:
tea -ek 'great elephant i' -I secret1 secret2 secret3
rm secret1 secret2 secret3
and this to decrypt
tea -dk 'great elephant i' -I secret1.3

Please advise if I'm doing something wrong or if there will be a fix for this.
Thanks
Will Hanson
willh@earthlink.net

**Comment by willhanson98** [comment]
cat of original file (test data):
secret1
this is what is in secret1
some uppercase ABC DEF GHI JKLM NOPQR STU VW XYZ
some numbers 123 456 789.12 -555.78
some symbols $&^%#@!~*()_+|

(there is a single newline at the end of each line in the original file that does not appear in this paste)

**Comment by willhanson98** [comment]
Test file contents (140 Bytes):
this is what is in secret1
some uppercase ABC DEF GHI JKLM NOPQR STU VW XYZ
some numbers 123 456 789.12 -555.78
some symbols $&^%#@!~*()_+|

Subsequent files end up being 144 bytes, looks like multiple of 8 bytes.

**Comment by coderarjob** [comment]
The code() function in `tea.c` encodes 8 bytes of data at a time. If in the end the file length is not a multiple of 8 bytes, zero is padded. The fix is quite easy, and will prevent these extra padding bytes to be written to the output file.

**Comment by coderarjob** [comment]
Change  `tea.c` line 107, ` if ((len = write (outf,d,DATA_SIZE)) < 0)` to `if ((len = write (outf,d,len)) < 0)`. Let me know this fixes the issue.

**Comment by willhanson98** [comment]
I tried that myself yesterday and it broke the ability to decrypt it properly. Please test with the 140 byte data I gave you and let me know.Thanks WillSent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: Arjob Mukherjee ***@***.***> Date: 4/19/21  4:05 AM  (GMT-06:00) To: coderarjob/tea-c ***@***.***> Cc: willhanson98 ***@***.***>, Author ***@***.***> Subject: Re: [coderarjob/tea-c] Using V 2.0.0 Text file in CentOS 7 do not get
  decrypted properly - extra 4 bytes at the end of the decrypted file (#3) 
Change  tea.c line 107,  if ((len = write (outf,d,DATA_SIZE)) < 0) to if ((len = write (outf,d,len)) < 0). Let me know this fixes the issue.

**Comment by willhanson98** [comment]
I tried your suggestion and while the encrypted filesize was correct, the decryption produced bad data near the end of the file. I tried to figure out how to fix that, but no luck. Please help. I would really like to use this tea method. Thanks, Will

**Comment by coderarjob** [comment]
Yes Will. The above suggested change, will break decryption logic. I will require some time to find a fix to this.

**Comment by willhanson98** [comment]
Thank you for spending the time on it. I appreciate it.

Wilnore (Will) Hanson
(770) 231-8309
***@***.***

**Comment by willhanson98** [comment]
Did you get a chance to work on it?

Wilnore (Will) Hanson
(770) 231-8309
***@***.***
